### DISCUSSION TITLE
nextui graph improvements

### DISCUSSION THREAD

**Discussion by hellt** [srl-labs/containerlab#818] [Ideas]
## Ideas and issues
- [ ] make link blocks with link names (e1-1) aligned with the link line to avoid overlaying the labels
- [ ] alternatively, make link labels to disappear with a click of a button. When link labels are not displayed, make them visible when hover over the node or a link element
- [x] make nextui a default template so that both `--template` and `--static-dir` can be omitted
- [ ] fix link tooltip to have a single line `source` and `destination` where source would be `leaf1:e1-1`.
- [ ] provide an option to define an icon per-node/kind. This can be done via [`extras`](https://github.com/srl-labs/containerlab/blob/b28babdde3c7db63a8aa0eb6886412291bf7b2bd/types/node_definition.go#L59) field that we have for NodeDefinition.
- [ ] allow to set custom color per node/group
- [ ] explore if export to PNG is possible
- [x] node tooltip doesn't fit long image URLs ([demo](https://gitlab.com/rdodin/pics/-/wikis/uploads/deeaf810acc9b33e297de230f961b1df/image.png)) [fixed in #834 ]
- [ ] add a field in the node tooltip that will open a web based ssh terminal to connect to the node directly

**Comment by hellt** [comment]
for "fix link tooltip to have a single line source and destination where source would be leaf1:e1-1" here is a prototype that can be used to achieve this

https://github.com/iDebugAll/devnet_marathon_endgame/commit/4aeb72cf045ecf160e1865e31425cfce4375dee3

**Comment by ahmedalodhari** [comment]
* [ ]  make link blocks with link names (e1-1) aligned with the link line to avoid overlaying the labels

Something like this?

<img width="980" alt="nextui-link-label" src="https://user-images.githubusercontent.com/11521160/160296314-dff143b4-db67-4d39-988d-37ae6973eb6c.png">

* [ ]  fix link tooltip to have a single line `source` and `destination` where source would be `leaf1:e1-1`.
* [ ]  explore if export to PNG is possible

Great Ideas. I'll look into them

* [ ]  provide an option to define an icon per-node/kind. This can be done via [`extras`](https://github.com/srl-labs/containerlab/blob/b28babdde3c7db63a8aa0eb6886412291bf7b2bd/types/node_definition.go#L59) field that we have for NodeDefinition.
* [ ]  allow to set custom color per node/group

Ideally, the graph options such as sortOrder, node icon, node color, etc. should be driven by the Clab configuration file. So, it would be nice if we can pass that info to the graph template. 

* [ ]  node tooltip doesn't fit long image URLs ([demo](https://gitlab.com/rdodin/pics/-/wikis/uploads/deeaf810acc9b33e297de230f961b1df/image.png))

Already fixed by #834

**Reply by hellt** [reply]
> Something like this?

yes, they can still be in this nice rectangular shapes with rounded corners, just aligned with the angle the link line has. This, potentially, will remove clutter when a node has several links located close to each other

**Comment by asadarafat** [comment]
- [ ] options to draw overlay services that can be part of the clab.
i made some prototype here, but struggling to make the links drawed on top of the group layer. 
@ahmedalodhari , maybe you coud give me some advice here to arrange elements Z-aligned...
https://github.com/asadarafat/containerlab/blob/main/templates/graph/nextui/static/js/script.js

![MicrosoftTeams-image](https://user-images.githubusercontent.com/19853888/160420453-c24bf555-2f25-4497-af30-b730851415f2.png)

**Reply by ahmedalodhari** [reply]
@asadarafat, since the z-index of the SVG elements is define by the order they appear in the DOM ([SVG 1.1 Spec](https://www.w3.org/TR/SVG11/render.html#RenderingOrder)), you have to reorder the `<g>` elements and put the `GroupsLayer` before the `LinksLayer`.

<img width="610" alt="nextui-svg-order" src="https://user-images.githubusercontent.com/11521160/160498534-755730cf-861c-4da6-9482-47141c1d25a1.png">

One way to do this is by using [Node.insertBefore()](https://developer.mozilla.org/en-US/docs/Web/API/Node/insertBefore):

```
var link_layer_el = document.querySelectorAll('[data-nx-type="nx.graphic.Topology.LinksLayer"]')
var group_layer_el = document.querySelectorAll('[data-nx-type="nx.graphic.Topology.GroupsLayer"]')
var parent_el = link_layer_el[0].parentNode
parent_el.insertBefore(group_layer_el[0], link_layer_el[0])
```
<img width="488" alt="nextui-reorder" src="https://user-images.githubusercontent.com/11521160/160501695-b2cb4bd8-6fef-4feb-a5a4-dcf049193291.png">

<img width="1075" alt="nextui-issue" src="https://user-images.githubusercontent.com/11521160/160502290-c915025c-a7c6-43ae-bc97-42bbadb7768e.png">

**Reply by asadarafat** [reply]
Thanks @ahmedalodhari  it works like magic 😄

**Reply by asadarafat** [reply]
Hi @ahmedalodhari, i would need your favor again to advice me on the following usecase:
supposed i have te following data/script https://github.com/asadarafat/containerlab/blob/main/templates/graph/nextui/static/js/script.js
i can get the following topology
![CleanShot 2022-06-18 at 21 36 43](https://user-images.githubusercontent.com/19853888/174454747-b5263bf2-d210-4281-a3e1-61b721578519.png)

however is seems the layout algorithm applied on all the node and it's makes the groups overlaped.
would it be possible to applied layout algorithm in selected nodes only (ie: L2-Layer) and then the nodes in L3-Layer will be posisioned  relative to the corresponding node in L2?
ie: supposed node L2-10.10.10.1 has posisition (x,y) then the node L3-10.10.10.3 position will be (x, y+50)

the target layout would be something like this
![CleanShot 2022-06-18 at 21 45 22](https://user-images.githubusercontent.com/19853888/174455006-13622c41-1859-4ada-86d5-89bc7808330c.png)


Thanks in advance

**Reply by ahmedalodhari** [reply]
@asadarafat, one way to avoid the overlapping group layers is to use the automatic hierarchical layout function to organize your nodes where the L3 nodes are ordered first and then the L2 nodes.
